cmd/clusterctl/cmd/generate_provider.go

```diff
@@ -0,0 +1,156 @@
+        # components defined in the provider repository.
+		clusterctl generate provider --infrastructure aws
+
+		# Generates a configuration file for creating provider for a specific version with variable values using 
+        # components defined in the provider repository.
+		clusterctl generate provider --infrastructure aws:v0.4.1
+
+		# Generates a raw configuration file for creating provider without variable values using 
+        # components defined in the provider repository.
+		clusterctl generate provider --infrastructure aws --raw
+
+		# Generates a raw configuration file for creating provider for a specific version without variable values using 
+        # components defined in the provider repository.
+		clusterctl generate provider --infrastructure aws:v0.4.1 --raw
+
+		# Prints list of variables used in the provider template.
+		clusterctl generate provider --list-variables
+
+		`),
+
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runGenerateProviderComponents()
+	},
+}
+
+func init() {
+	generateProviderCmd.Flags().StringVar(&gpo.coreProvider, "core", "",
+		"Core provider and version (e.g. cluster-api:v0.3.0)")
+	generateProviderCmd.Flags().StringVarP(&gpo.infrastructureProvider, "infrastructure", "i", "",
+		"Infrastructure provider and version (e.g. aws:v0.5.0)")
+	generateProviderCmd.Flags().StringVarP(&gpo.bootstrapProvider, "bootstrap", "b", "",
+		"Bootstrap provider and version (e.g. kubeadm:v0.3.0)")
+	generateProviderCmd.Flags().StringVarP(&gpo.controlPlaneProvider, "control-plane", "c", "",
+		"ControlPlane provider and version (e.g. kubeadm:v0.3.0)")
+	generateProviderCmd.Flags().StringVar(&gpo.targetNamespace, "target-namespace", "",
+		"The target namespace where the provider should be deployed. If unspecified, the components default namespace is used.")
+	generateProviderCmd.Flags().StringVar(&gpo.watchingNamespace, "watching-namespace", "",
+		"Namespace the provider should watch when reconciling objects. If unspecified, all namespaces are watched.")
+	generateProviderCmd.Flags().BoolVar(&gpo.skipVariables, "raw", false,
+		"Generate configuration without variable substitution.")
+	generateProviderCmd.Flags().BoolVar(&gpo.listVariables, "list-variables", false,
+		"Returns the list of variables expected by the provider template")
+
+	generateCmd.AddCommand(generateProviderCmd)
+}
+
+func runGenerateProviderComponents() error {
+	providerName, providerType, err := parseProvider()
+	if err != nil {
+		return err
+	}
+	c, err := client.New(cfgFile)
+	if err != nil {
+		return err
+	}
+
+	options := client.ComponentsOptions{
+		TargetNamespace:   gpo.targetNamespace,
+		WatchingNamespace: gpo.watchingNamespace,
+		SkipVariables:     gpo.skipVariables,
+	}
+	components, err := c.GetProviderComponents(providerName, providerType, options)
+	if err != nil {
+		return err
+	}
+
+	if gpo.listVariables {
+		return printVariablesOutput(components)
+	}
+
+	return printYamlOutput(components)
```

Comment:
So, the `describe` view was the default functionality for `clusterctl config provider`, and it didn't make sense to do that for `clusterctl generate provider`. `--list-variables` kind of seemed consistent with other generate subcommands. Having said that describe also lists a few more things (like images) in addition to variables. Do you think I should add a flag for `--describe`?

Reply:
I don't have strong opinions here; given that we are moving to `clusterctl generate provider` I makes sense to me to have yaml as a default format and other variants to be triggered with flags; might be we should --show-summary instead of --describe (it is a little bit more explicit about the intent).
I'll defer the last call to @CecileRobertMichon, which If I remember well is one of the users relying on the current output of `clusterctl config provider`...